### PR TITLE
#60: Use getEXAType() to determin Exasol-specific types

### LINKED ISSUE #60: Use getEXAType() to determin Exasol-specific types
## Situation

Exasol has uncommon types like `INTERVAL`, `HASHTYPE` and `GEOMETERY`. While types are like that are not unheard of in other products, they are rare. For compatibility with 3rd party products, the JDBC driver's `getType()` method reports them as either `VARCHAR` or `CHAR`. This is a problem in the Exasol Virtual Schema though, because you really want the exact original types in a connection between two Exasol clusters.

Luckily the driver has a non-standard `getEXAType()` method that reports the original type.

Back reference: SUPPORT-27418

## Breaking Change

While this fixes a long lasting problem, it is still a breaking change, so we need to write that into the release letter and increase the major version. Same for RLS when we propagate this.

## Acceptance Criteria

1. Metadata reader determines original column types for all columns.

### PR DESCRIPTION
**Pull request by kaklakariada**
This improves type mapping for local and EXA connections as described in #60.
Correct type mapping for JDBC connections requires a new JDBC driver, see https://github.com/exasol/exasol-virtual-schema/issues/60#issuecomment-977765705

### CODE REVIEW

File: doc/changes/changes_6.0.0.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Exasol Virtual Schema 6.0.0, released 2021-12-??
+
+Code name: Support native Exasol data types
```
Comment (jakobbraun) [inline doc/changes/changes_6.0.0.md:3]: Maybe add a note that we should not release before the JDBC driver is released. If you add it to the first line, everyone will read it before updating the date.

File: doc/changes/changes_6.0.0.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+This release improves support for Exasol's data types `HASHTYPE`, `GEOMETRY`, `INTERVAL DAY TO SECOND`, and `INTERVAL YEAR TO MONTH`. The virtual schema now reports them as their real type, not as `CHAR` or `VARCHAR` as before. **This is a breaking change!**
```
Comment (jakobbraun) [inline doc/changes/changes_6.0.0.md:5]:
```suggestion
## Summary

This release improves support for Exasol's data types `HASHTYPE`, `GEOMETRY`, `INTERVAL DAY TO SECOND`, and `INTERVAL YEAR TO MONTH`. The virtual schema now reports them as their real type, not as `CHAR` or `VARCHAR` as before. **This is a breaking change!**
```

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java [hunk 3]
```diff
@@ -37,22 +53,102 @@ public ExasolColumnMetadataReader(final Connection connection, final AdapterProp
...
+    private DataType map(final JDBCTypeDescription jdbcTypeDescription) {
+        DataType type = createTypeBasedOnTypeName(jdbcTypeDescription);
+        if (type != null) {
+            return type;
+        }
+        type = createTypeBasedOnJdbcType(jdbcTypeDescription);
+        if (type != null) {
+            return type;
+        }
+
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java:73]:
```suggestion
```

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java [hunk 4]
```diff
@@ -37,22 +53,102 @@ public ExasolColumnMetadataReader(final Connection connection, final AdapterProp
...
+    }
+
+    private DataType createTypeBasedOnJdbcType(final JDBCTypeDescription jdbcTypeDescription) {
         switch (jdbcTypeDescription.getJdbcType()) {
-        case EXASOL_INTERVAL_DAY_TO_SECONDS:
-            return DataType.createIntervalDaySecond(jdbcTypeDescription.getPrecisionOrSize(),
-                    jdbcTypeDescription.getDecimalScale());
-        case EXASOL_INTERVAL_YEAR_TO_MONTHS:
-            return DataType.createIntervalYearMonth(jdbcTypeDescription.getPrecisionOrSize());
+        case EXASOL_HASHTYPE:
```
Comment (jakobbraun) [inline src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java:101]: The issue sais you wanted to use `getEXAType` but I could not find it in the PR. Is that intended?

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java (hunk 4 above)
Comment (kaklakariada) [reply src/main/java/com/exasol/adapter/dialects/exasol/ExasolColumnMetadataReader.java:100]: Yes. getEXAType() is not yet accessible in JDBC driver 7.1.3. We need to wait for 7.1.4. Only then we can really fix #60

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**